### ISSUE TITLE
Data not outputting correctly

### ISSUE THREAD

**Issue by tpats98** [TravisLedo/LidarScanRender#3]
Hi Travis,

I am a graduate student working on a project to eventually be able to scan and map underwater interiors such as caves and wrecks. I found your project and have been attempting to replicate it in my own way. I'm using Arduino to control my servo/LiDAR system as well as perform the conversion calculations. The sensor I am using is a Garmin LiDAR Lite v3HP. I use Putty to put the data into an Excel file and I am then using a Matlab code to put that data into a point cloud format. So, it is not happening live as in your scan. I am running into issues where the point cloud that is created looks nothing like the room I scanned. One of the issues is I am not getting any straight edges as if it was scanning a wall. It seems as though it is giving me a circle just based on the cloud I am seeing. The point clouds just do not make sense compared to what it should be scanning. I attempted to attach a screenshot, I am not sure if it is visible. Overall, it seems that my inputs are good. I am getting my pan and tilt servo positions as well as a distance reading from my sensor. This makes me think the issue is somewhere in my calculations and post-processing or maybe it could be an issue with my sensor/setup itself. 

I am new to Github so I wasn't sure how to reach you. My apologies if this isn't the appropriate way to contact you. I was hoping you might be willing to provide some aid to me as I try to solve my issues. I look forward to hearing from you!

Thank you,
Tim Patschorke

![Screenshot (51)](https://user-images.githubusercontent.com/72276194/228023716-b8399cbc-d019-4702-a12d-2c5f1290c9b3.png)

**Comment by TravisLedo** [comment]
Hey Tim,

Thanks for taking interest in this project and your app is looking really good. 

When making your points render on screen, did you account for all the conversions? You have to convert from spherical coordinates, to rectangular coordinates, and then also to computer screen coordinates as well. Screens have minor differences like the Y axis going down for positive, not up. I would start a simple scan with just a few points first where you kind of know what the results should be and then play around with the conversions until it looks right. Even better if you test with just only one servo at a time to eliminate one axis from the conversions.

Another thing to make sure is that you are not using an Arduino UNO, the MEGA is a better choice. This project requires more than one serial line. One for sending data to pc and also another for getting data from the lidar sensor to the Arduino.  They interfere with each other if there is only one and the software serial libraries also will not work.

I am sure you have done this but have u also tried just testing the sensor itself? Just moving your hand back and forth to make sure the distance output looks correct.

Hope one of these could help you solve the problem. Let me know how it goes. Good luck!

**Comment by tpats98** [comment]
Hi Travis,

Thank you for the timely response. I am mostly accounting for the conversions. I am having my Arduino code perform the spherical to cartesian as it scans. I have done a little bit of tinkering with the conversion for the computer coordinate system but not as much. 

I tried taking out an axis, as you mention, by only using the pan servo. I played around with this and different things over the weekend and the only thing that would make my cloud look remotely close was by creating my own range for x. Instead of using the data, I just created a 0-180 range that the distance value could be plotted against. This gave me a somewhat accurate depiction of what was in front of the sensor but it is not a true representation of the room obviously. 

As for the board, I am using a Mega. I haven't run into any issue whatsoever in the actual operation of the servos and sensor so that is a plus haha.

As for the sensor, you are correct, I have tested it by itself. I did not see any problems jumping out at me. 

Overall, do you think my issue is just in the conversions possibly? I guess my main confusion is why it is outputting as a half circle instead of straight lines as if it was scanning a wall or came to a corner. Any ideas on that other than conversion issues maybe?

**Comment by TravisLedo** [comment]
Yea I am leaning towards a conversion issue, especially when it looks more spherical when plotted. I would try to send manual data from the Arduino into the app as well to see if you get what you expect on the screen. Start with with something simple like (0,0,0,) and (10,10,10) and see if the two points are plotting correctly then add more points to the test.  If this looks good then most likely a conversion issue.  

It could maybe be the lidar scanner itself not able to handle really narrow/wide bounces well. Try reducing the width of the scan and see if the render looks more like the room and less rounded. The roundness seems to happen more on the edges, so that could be it.

Have you tried dramatically slowing down the servos' delay between each point to give it more time to move to position, scan, move to the next position? Did you scan one point multiple times and play around with using the average or median? Try to get random readings and see how consistent the scanner is at one position.

Also have you tried a different wall? lol. I know that lidar sucks on glossy surfaces as well as sharp corners.

**Comment by tpats98** [comment]
I will troubleshoot and try those suggestions as I keep working today and tomorrow. Hopefully with enough tinkering of the conversions and any adjustments, I will start to see what I should see.

**Comment by tpats98** [comment]
Hey Travis,

I wanted to reach out again as  I have good news! I spent the last 2 days messing around with the conversions and it is looking like I am getting the scan I need. I have attached one in case you're interested. I just wanted to thank you again for pointing me in the right direction! 
![Screenshot (52)](https://user-images.githubusercontent.com/72276194/228616846-a2d2ada6-657e-4ba1-99c8-bc50a9503850.png)
